// publish-swagger-model/api-model/index.js
/**
 * Created by rajasekhar on 01-Jul-17.
 */
//example of api-model object,  so clinet should request model like ths
// model = {
//     isSandboxAvailable : 'true',
//     swaggerurl : "http://qa-assist24.app.shared.int.sv2.247-inc.net:8080/org-hierarchy-service/api/swagger-json",
//     proxyurl : "http://dev.api.sv2.247-inc.net/ohs",
//     createdBy : "test11111",
//     menu : "test",
//     parentMenu : "test",
//     approver : "test"
//
// };
const logger = require('logger')
const  publishRecordColumns = require('publish-record-columns');
class APIModel {
    
}

APIModel.prototype.model_name = 'name'
APIModel.prototype.model_id = 'model_id'
APIModel.prototype.latest_revision_number = 'latest_revision_number'
APIModel.prototype.published_revision_number = 'published_revision_number'
APIModel.prototype.isDeleted = 'isDeleted'


APIModel.prototype.convertReqToApiModel = function (req) {
    var model = undefined;
    if (req && req.body && req.body.data && req.body.data.model) {
        model = req.body.data.model;
    }
    var keys = [publishRecordColumns.isSandBoxAvailable, publishRecordColumns.swagger_url,   publishRecordColumns.proxyurl,  publishRecordColumns.product,  publishRecordColumns.reviewer];
    for (var index in keys) {
        if (model[keys[index]] === '' || model[keys[index]] == undefined) {
            var message = 'requested model either does not have property : '+keys[index] +' or  it is undifined '
            logger.error(message)
            return {message : message, isValid : false};
        }
    }
    logger.info('requested model is valid')
    return model;
}
module.exports = new APIModel();
